R/list_of.R

```diff
@@ -120,6 +120,13 @@ as.list.vctrs_list_of <- function(x, ...) {
 
 #' @export
 `[[<-.vctrs_list_of` <- function(x, i, value) {
+  if (is.null(value)) {
+    # Setting to NULL via [[ shortens the list! Example:
```

Comment:
I think if we're going to do this  we should also add a check for `rlang::zap()`, so there's still some way to remove elements.

Reply:
Currently:

``` r
library(vctrs)

x <- list_of(1, 2, 3)
x[[2]] <- rlang::zap()
#> Error: Can't cast <rlang_zap> to <double>
#> Backtrace:
#>     █
#>  1. ├─base::`[[<-`(`*tmp*`, 2, value = structure(list(), class = "rlang_zap"))
#>  2. └─vctrs:::`[[<-.vctrs_list_of`(`*tmp*`, 2, value = structure(list(), class = "rlang_zap"))
#>  3.   ├─vctrs::vec_cast(value, attr(x, "ptype")) /home/kirill/git/R/vctrs/R/list_of.R:130:2
#>  4.   ├─vctrs::vec_cast.double(value, attr(x, "ptype")) /home/kirill/git/R/vctrs/R/cast.R:102:2
#>  5.   └─vctrs:::vec_cast.double.default(value, attr(x, "ptype")) /home/kirill/git/R/vctrs/R/cast.R:219:2
#>  6.     └─vctrs::stop_incompatible_cast(x, to) /home/kirill/git/R/vctrs/R/cast.R:251:2
#>  7.       └─vctrs:::stop_incompatible(...) /home/kirill/git/R/vctrs/R/conditions.R:53:2
```

<sup>Created on 2019-01-29 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1.9000)</sup>

Do we need to add a special type, akin to `unspecified` ?

Reply:
No, just special case it for this usage.

Reply:
Added, but do we really need it? How do we remove elements in an integer vector?

Reply:
We don't worry about it.

Reply:
> so there's still some way to remove elements.

So why do we worry about removing elements for list vectors?

I don't mind either way, just being picky here because I think consistency is important:

- Setting elements to `NA` should be supported (`NULL` for lists)
- Removing elements should be either supported consistently, or not supported at all

Reply:
I've changed my mind on this after talking to @lionel-. I think we should ensure that ``vec_size(`[[<-`(x, i, value)`` equals `vec_size(x)`, so that we don't need to support `zap()` here.

Reply:
I've removed special-casing for `zap` here.

The `vec_size()` still can change when creating new elements past the existing length.